esp/services/ws_decoupledlogging/ws_decoupledloggingservice.cpp

```diff
@@ -0,0 +1,165 @@
+    limitations under the License.
+############################################################################## */
+
+#include "ws_decoupledloggingservice.hpp"
+
+#include "jlib.hpp"
+
+#include "LoggingErrors.hpp"
+#include "loggingcommon.hpp"
+#include "loggingagentbase.hpp"
+#include "exception_util.hpp"
+
+typedef IEspLogAgent* (*newLoggingAgent_t_)();
+
+IEspLogAgent* CWSDecoupledLogEx::loadLoggingAgent(const char* name, const char* dll, const char* service, IPropertyTree* cfg)
+{
+    StringBuffer plugin;
+    plugin.append(SharedObjectPrefix).append(dll).append(SharedObjectExtension);
+    HINSTANCE loggingAgentLib = LoadSharedObject(plugin.str(), true, false);
+    if(!loggingAgentLib)
+        throw MakeStringException(EspLoggingErrors::LoadLoggingLibraryError, "can't load library %s", plugin.str());
+
+    newLoggingAgent_t_ xproc = (newLoggingAgent_t_)GetSharedProcedure(loggingAgentLib, "newLoggingAgent");
+    if (!xproc)
+        throw MakeStringException(EspLoggingErrors::LoadLoggingLibraryError, "procedure newLoggingAgent of %s can't be loaded", plugin.str());
+
+    return (IEspLogAgent*) xproc();
+}
+
+void CWSDecoupledLogEx::init(IPropertyTree* cfg, const char* process, const char* service)
+{
+    if (!cfg)
+        throw MakeStringException(-1, "Can't initialize CWSDecoupledLogEx, cfg is NULL");
+
+    espProcess.set(process);
+
+    StringBuffer xpath;
+    xpath.setf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
+    IPropertyTree *serviceCFG = cfg->queryPropTree(xpath.str());
+    tankFileDir.set(serviceCFG->queryProp("TankFileDir"));
+    if (!tankFileDir.length())
+        throw MakeStringException(-1, "Can't initialize CWSDecoupledLogEx, TankFileDir is NULL");
+
+    Owned<IPTreeIterator> loggingAgentSettings = serviceCFG->getElements("LogAgent");
+    ForEach(*loggingAgentSettings)
+    {
+        IPropertyTree& loggingAgentTree = loggingAgentSettings->query();
+        const char* agentName = loggingAgentTree.queryProp("@name");
+        const char* agentType = loggingAgentTree.queryProp("@type");
+        const char* agentPlugin = loggingAgentTree.queryProp("@plugin");
+        if (!agentName || !*agentName || !agentPlugin || !*agentPlugin)
+            continue;
+
+        IEspLogAgent* loggingAgent = loadLoggingAgent(agentName, agentPlugin, service, cfg);
+        if (!loggingAgent)
+        {
+            OERRLOG(-1, "Failed to create logging agent for %s", agentName);
+            continue;
+        }
+        loggingAgent->init(agentName, agentType, &loggingAgentTree, service);
+        loggingAgent->initVariants(&loggingAgentTree);
+        IUpdateLogThread* logThread = createUpdateLogThread(&loggingAgentTree, service, agentName, tankFileDir.get(), loggingAgent);
+        if(!logThread)
+            throw MakeStringException(-1, "Failed to create update log thread for %s", agentName);
+        loggingAgentThreads.push_back(logThread);
```

Comment:
Should insertion be conditional on having a log request reader available? That would eliminate the need to throw exceptions when an unsupported agent is specified in any service method.

Reply:
Ok